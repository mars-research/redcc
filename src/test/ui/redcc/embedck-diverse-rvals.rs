// build-pass

#![feature(rustc_attrs)]

#[cfg_attr(not(test), rustc_diagnostic_item = "RRef")]
struct RRef<T>(T);
struct Direct(RRef<i32>);
struct Ref<'a>(&'a RRef<i32>);

enum MaybeRRef {
    Yes(RRef<i32>),
    No,
}

fn main() {
    // array
    let mut array = RRef([RRef(0), RRef(1)]);
    array.0 = [RRef(1), RRef(2)];

    // tuple
    let mut tuple = RRef((RRef(0), RRef(1)));
    tuple.0 = (RRef(0), RRef(1));

    // reference
    let tmp = &RRef(1);
    let mut r = RRef(Ref(tmp));
    r.0.0 = &RRef(0);

    // enum
    let mut m = RRef(MaybeRRef::No);
    m.0 = MaybeRRef::Yes(RRef(1));

    // raw pointer
    let p: *const RRef<_> = &RRef(1);
    let mut r = RRef(p);
    r.0 = &RRef(2);
}
