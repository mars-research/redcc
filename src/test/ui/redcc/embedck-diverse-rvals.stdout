EMBED [RRef<i32>; 2] -> RRef<[RRef<i32>; 2]>
EMBED (RRef<i32>, RRef<i32>) -> RRef<(RRef<i32>, RRef<i32>)>
EMBED &RRef<i32> -> RRef<Ref>
EMBED MaybeRRef -> RRef<MaybeRRef>
EMBED *const RRef<i32> -> RRef<*const RRef<i32>>
